verification/certifyKey_test.go

```diff
@@ -124,15 +423,210 @@ func testCertifyKey(d TestDPEInstance, t *testing.T) {
+
+	// Run X.509 linter on full certificate chain and file issues for errors
+	leafCert := checkCertificateStructure(t, leafCertBytes)
+	if leafCert == nil {
+		t.Fatalf("Could not obtain leaf certificate from CertifyKey command.")
+	}
+	t.Logf("Leaf certificate is DER encoded")
+
+	certChain := checkCertificateChain(t, certChainBytes)
+	if certChain == nil || len(certChain) == 0 {
+		t.Fatalf("Could not obtain certificate chain from GetCertificateChain command.")
+	}
+	t.Logf("Certificate chain is DER encoded")
+
+	// Validate that all X.509 fields conform with the format defined in the DPE iRoT profile
+	validateCertifyKeyFields(t, leafCert)
+
+	// Validate that proper TcbInfo extensions are populated with expected measurements and values as defined in the DPE iRoT profile
+
+	// Validate correct KeyUsage and BasicConstraints extensions for all DICE certificates
+
+	// Ensure full certificate chain has valid signatures
+	validateCertChain(t, certChain, leafCert)
 
 	// TODO: When DeriveChild is implemented, call it here to add more TCIs and call CertifyKey again.
 }
+
+// Validate signature of certificates in certificate chain passed.
+func validateSignature(t *testing.T, certchain []*x509.Certificate) {
+	t.Helper()
+	foundError := false
+	for i := 0; i < len(certchain)-1; i++ {
+		err := certchain[i].CheckSignatureFrom(certchain[i+1])
+		if err != nil {
+			t.Logf("Signature validation failed for certificate %s in chain with error %s", certchain[i].Subject.String(), err.Error())
+			foundError = true
+		} else {
+			t.Logf("Signature validation passed %s", certchain[i].Subject.String())
+		}
+
+	}
+
+	root := certchain[len(certchain)-1]
+	//err := root.CheckSignatureFrom(root)
+	err := root.CheckSignature(root.SignatureAlgorithm, root.RawTBSCertificate, root.Signature)
+	if err != nil {
+		t.Logf("Root certificate  %s Signature validation failed with error %s", root.Subject.String(), err.Error())
+		foundError = true
+	}
+	t.Logf("Root certificate Signature validation passed %s\n", root.Subject.String())
+
+	if foundError {
+		// TODO: Fail the test when the certificate is expected to pass
+		t.Logf("Found certificates with mismatching signature")
+	}
+}
+
+// Build certificate chain and calls to validateSignature on each chain.
+func validateCertChain(t *testing.T, certChain []*x509.Certificate, leafCert *x509.Certificate) {
+	t.Helper()
+	if leafCert != nil {
+		// Remove unhandled critical extensions reported by x509 but defined in spec
+		t.Log("Validating leaf certificate chain...")
+		t.Log("Checking for unhandled critical certificate extensions unknown to DPE certificates profile spec...")
+		removeTcgDiceCriticalExtensions(t, []*x509.Certificate{leafCert})
+
+		// Build verify options
+		opts := buildVerifyOptions(certChain)
+
+		// Certificate chain validation for leaf
+		chains, err := leafCert.Verify(opts)
+		if err != nil {
+			// TODO: Fail the test with Errorf here once we expect it to pass.
```

Comment:
Yes updated comments.
Here are the consolidated list of errors ignored for now, will be enabled when certs are received proper format.

1. Signature validation of leaf certificate fails
2. Certificate chain cannot be built from leaf to root
3. When DeriveChild is implemented, call it here to add more TCIs and call CertifyKey again.
4. multiTcb info is not provided in leaf
5. Linter error
6. Ueid extn value doen not match the label	(as per irot spec saying "UEID shall be populated by "label"  ")

Reply:
Reg 4; the spec says it would return array of DiceTcbInfo, but we receive single DiceTcbInfo, is this expected?

Reg 6: This SUCCEEDS if processed as per your comment, I just compared Label.Equals(UeidExtnValue), 

Reg 7: Yes, when ADDISCA is set 1, isCA is still false

> 
> 	certifyKeyReq := []CertifyKeyReq[SHA256Digest]{
> 		{
> 			ContextHandle: [16]byte{0},
> 			Flags:         0,
> 			Label:         [32]byte{},
> 			Format:        CertifyKeyX509,
> 		},
> 		{
> 			ContextHandle: [16]byte{0},
> 			Flags:         1073741824, // Set the 31st bit: 01000000 00000000 00000000 00000000
> 			Label:         [32]byte{},
> 			Format:        CertifyKeyX509,
> 		},
> 	}

	
For nonzero flags also, I receive only isCAFalse, so the test fails. I have enabled error, you may notice this is go test.

Reply:
Item 7 related reference : https://github.com/hpya93/caliptra-dpe/blob/main/verification/certifyKey_test.go#L371

Reply:
**Reg your question  -- whether the leaf cert validation succeeds now?**
No, because key usage in cert chain is empty - not listed and this causes leaf to fail with "Invalid key usage"

Please find the decoded certs I collected from "go test --verbose".

```
haripriya@haripriyab:~$ openssl x509 -inform PEM -in k.pem -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            62:30:37:30:65:64:36:31:39:39:39:31:39:38:39:39:38:33:62:63
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: C = US, ST = Illinois, L = Chicago, O = Google, OU = PSS, CN = Sree, emailAddress = srrevoori@google.com
        Validity
            Not Before: Feb 27 00:00:00 2023 GMT
            Not After : Dec 31 23:59:59 9999 GMT
        Subject: CN = DPE Leaf, serialNumber = b070ed619991989983bc305ff9beb69c67e9e8895ed5417f846af4cf483dc120
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:29:78:6b:1a:85:23:9b:53:e4:e7:22:4b:2b:35:
                    87:68:e9:06:a6:ee:3c:12:28:64:32:83:c0:cf:6a:
                    70:bf:ed:1a:a8:b5:50:ce:e9:22:c7:08:99:79:d4:
                    0f:37:e4:15:92:d9:c3:53:1b:8e:67:20:c9:9f:e4:
                    92:b9:d5:d8:3d
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            2.23.133.5.4.5: critical
                0n0l.^0-..`.H.e..... ................................0-..`.H.e..... ............................................
            2.23.133.5.4.4: critical
                0". ................................
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Key Usage: critical
                Digital Signature
            X509v3 Extended Key Usage: critical
                2.23.133.5.4.100.9
    Signature Algorithm: ecdsa-with-SHA256
    Signature Value:
        30:45:02:20:39:a0:b8:d3:e3:e2:88:9d:e5:74:e7:89:ef:36:
        a9:81:c6:d3:f2:5f:5b:d5:ea:21:dd:3b:c7:39:b4:73:e2:29:
        02:21:00:bd:f1:1c:c3:41:3b:13:1f:38:ea:7e:d7:f5:c9:0b:
        80:7d:8c:c3:ca:1d:88:48:7e:1c:ea:49:00:c6:ff:30:9a
-----BEGIN CERTIFICATE-----
MIIC2TCCAn+gAwIBAgIUYjA3MGVkNjE5OTkxOTg5OTgzYmMwCgYIKoZIzj0EAwIw
gYUxCzAJBgNVBAYTAlVTMREwDwYDVQQIDAhJbGxpbm9pczEQMA4GA1UEBwwHQ2hp
Y2FnbzEPMA0GA1UECgwGR29vZ2xlMQwwCgYDVQQLDANQU1MxDTALBgNVBAMMBFNy
ZWUxIzAhBgkqhkiG9w0BCQEWFHNycmV2b29yaUBnb29nbGUuY29tMCIYDzIwMjMw
MjI3MDAwMDAwWhgPOTk5OTEyMzEyMzU5NTlaMF4xETAPBgNVBAMTCERQRSBMZWFm
MUkwRwYDVQQFE0BiMDcwZWQ2MTk5OTE5ODk5ODNiYzMwNWZmOWJlYjY5YzY3ZTll
ODg5NWVkNTQxN2Y4NDZhZjRjZjQ4M2RjMTIwMFkwEwYHKoZIzj0CAQYIKoZIzj0D
AQcDQgAEKXhrGoUjm1Pk5yJLKzWHaOkGpu48EihkMoPAz2pwv+0aqLVQzukixwiZ
edQPN+QVktnDUxuOZyDJn+SSudXYPaOB7jCB6zB9BgZngQUFBAUBAf8EcDBuMGym
XjAtBglghkgBZQMEAgEEIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
MC0GCWCGSAFlAwQCAQQgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACI
BAAAAACJBAAAAAAwMQYGZ4EFBQQEAQH/BCQwIgQgAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAwDwYDVR0TAQH/BAUwAwEBADAPBgNVHQ8BAf8EBQMDAIAB
MBUGA1UdJQEB/wQLMAkGB2eBBQUEZAkwCgYIKoZIzj0EAwIDSAAwRQIgOaC40+Pi
iJ3ldOeJ7zapgcbT8l9b1eoh3TvHObRz4ikCIQC98RzDQTsTHzjqftf1yQuAfYzD
yh2ISH4c6kkAxv8wmg==
-----END CERTIFICATE-----
haripriya@haripriyab:~$ openssl x509 -inform PEM -in l.pem -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            44:e2:18:b7:a0:92:51:0a:a9:3b:8e:fc:c0:82:be:4d:6c:ca:24:b4
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: C = US, ST = Illinois, L = Chicago, O = Google, OU = PSS, CN = Sree, emailAddress = srrevoori@google.com
        Validity
            Not Before: May 31 23:43:29 2023 GMT
            Not After : May 30 23:43:29 2024 GMT
        Subject: C = US, ST = Illinois, L = Chicago, O = Google, OU = PSS, CN = Sree, emailAddress = srrevoori@google.com
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:4e:0d:2f:e1:79:c3:ea:97:07:58:6f:d4:78:b7:
                    15:01:df:1a:3c:ea:f7:7b:9f:fe:b8:a6:77:b1:3c:
                    c7:21:34:3c:e6:96:16:5b:b3:7b:5c:82:76:cf:39:
                    93:50:ef:d6:ee:df:4d:49:b0:8e:fb:e7:56:37:2b:
                    2b:ed:7f:a1:80
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                69:6F:35:61:11:12:7B:C6:90:BD:E4:42:4C:29:4B:80:2C:E9:37:44
            X509v3 Authority Key Identifier: 
                69:6F:35:61:11:12:7B:C6:90:BD:E4:42:4C:29:4B:80:2C:E9:37:44
            X509v3 Basic Constraints: critical
                CA:TRUE
    Signature Algorithm: ecdsa-with-SHA256
    Signature Value:
        30:45:02:20:63:e9:7d:94:41:91:65:a8:ab:55:81:30:8d:b9:
        c7:16:5c:09:86:d5:c1:eb:01:c6:cb:01:0c:d9:25:43:76:57:
        02:21:00:84:0d:b5:96:8e:32:64:b4:b5:5f:39:00:7d:07:a3:
        ec:e2:56:71:8b:97:32:c0:69:9b:b4:b8:5a:64:d7:e1:9a
-----BEGIN CERTIFICATE-----
MIICYTCCAgegAwIBAgIUROIYt6CSUQqpO478wIK+TWzKJLQwCgYIKoZIzj0EAwIw
gYUxCzAJBgNVBAYTAlVTMREwDwYDVQQIDAhJbGxpbm9pczEQMA4GA1UEBwwHQ2hp
Y2FnbzEPMA0GA1UECgwGR29vZ2xlMQwwCgYDVQQLDANQU1MxDTALBgNVBAMMBFNy
ZWUxIzAhBgkqhkiG9w0BCQEWFHNycmV2b29yaUBnb29nbGUuY29tMB4XDTIzMDUz
MTIzNDMyOVoXDTI0MDUzMDIzNDMyOVowgYUxCzAJBgNVBAYTAlVTMREwDwYDVQQI
DAhJbGxpbm9pczEQMA4GA1UEBwwHQ2hpY2FnbzEPMA0GA1UECgwGR29vZ2xlMQww
CgYDVQQLDANQU1MxDTALBgNVBAMMBFNyZWUxIzAhBgkqhkiG9w0BCQEWFHNycmV2
b29yaUBnb29nbGUuY29tMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAETg0v4XnD
6pcHWG/UeLcVAd8aPOr3e5/+uKZ3sTzHITQ85pYWW7N7XIJ2zzmTUO/W7t9NSbCO
++dWNysr7X+hgKNTMFEwHQYDVR0OBBYEFGlvNWEREnvGkL3kQkwpS4As6TdEMB8G
A1UdIwQYMBaAFGlvNWEREnvGkL3kQkwpS4As6TdEMA8GA1UdEwEB/wQFMAMBAf8w
CgYIKoZIzj0EAwIDSAAwRQIgY+l9lEGRZairVYEwjbnHFlwJhtXB6wHGywEM2SVD
dlcCIQCEDbWWjjJktLVfOQB9B6Ps4lZxi5cywGmbtLhaZNfhmg==
-----END CERTIFICATE-----

```